poetry/installation/executor.py

```diff
@@ -107,14 +107,30 @@ def execute(self, operations):  # type: (Operation) -> int
         self._sections = OrderedDict()
         for _, group in groups:
             tasks = []
+            serial_operations = []
             for operation in group:
                 if self._shutdown:
                     break
 
+                # Some operations are unsafe, we mus execute them serially in a group
+                # https://github.com/python-poetry/poetry/issues/3086
+                # https://github.com/python-poetry/poetry/issues/2658
+                is_parallel_unsafe = operation.job_type == "uninstall" or (
+                    operation.package.develop
+                    and operation.package.source_type == "directory"
+                )
+                if not operation.skipped and is_parallel_unsafe:
+                    serial_operations.append(operation)
+                    continue
+
                 tasks.append(self._executor.submit(self._execute_operation, operation))
 
             try:
                 wait(tasks)
+
+                for operation in serial_operations:
+                    wait([self._executor.submit(self._execute_operation, operation)])
```

Comment:
Don't have to submit and wait here. Can just call `_execute_operation`.

Reply:
Yup, but I wanted keep calls consistent via `self._executor` for now.